core/trino-main/src/test/java/io/trino/execution/TestCreateTableColumnTypeCoercion.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.connector.MockConnectorFactory;
+import io.trino.spi.type.TimestampType;
+import io.trino.testing.LocalQueryRunner;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static io.airlift.testing.Closeables.closeAllRuntimeException;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static io.trino.testing.assertions.TrinoExceptionAssert.assertTrinoExceptionThrownBy;
+
+public class TestCreateTableColumnTypeCoercion
+{
+    private static final String catalogName = "mock";
+    private LocalQueryRunner queryRunner;
+
+    private LocalQueryRunner createLocalQueryRunner()
+    {
+        Session session = testSessionBuilder()
+                .setCatalog(catalogName)
+                .setSchema("default")
+                .build();
+        LocalQueryRunner queryRunner = LocalQueryRunner.create(session);
+        queryRunner.createCatalog(
+                catalogName,
+                prepareConnectorFactory(catalogName),
+                ImmutableMap.of());
+        return queryRunner;
+    }
+
+    private MockConnectorFactory prepareConnectorFactory(String catalogName)
+    {
+        return MockConnectorFactory.builder()
+                .withName(catalogName)
+                .withGetTableHandle(((session, schemaTableName) -> null))
+                .withCoerceNewTableColumn((session, type) -> {
+                    if (type instanceof TimestampType) {
+                        return Optional.of(VARCHAR);
+                    }
+                    return Optional.empty();
+                })
+                .build();
+    }
+
+    @Test
+    public void testIncompatibleType()
+    {
+        assertTrinoExceptionThrownBy(() -> queryRunner.execute("CREATE TABLE test_incompatible_type AS SELECT TIMESTAMP '2020-09-27 12:34:56.999' a"))
```

Comment:
That won't be possible through this class, getting an [exception](https://github.com/trinodb/trino/blob/2d41a082f0c08ea3ae02628a5b640127881006b5/core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java#L384).

This class is mainly created to test the interaction with the MetadataManager which is actually the same for CTAS, CT or CTL.

The TestCreateTableTask tests the CT and CTL more in depth but these are not covered by integration tests in trino-main. I could extend the testing in this file by adding more mock methods but it would be only for CTAS.

Reply:
CTAS with no data ?